src/main/java/net/oauth/jsontoken/AbstractJsonTokenParser.java

```diff
@@ -0,0 +1,229 @@
+   *        secret keys.
+   */
+  public boolean signatureIsValid(String tokenString, List<Verifier> verifiers) {
+    String[] pieces = splitTokenString(tokenString);
+    byte[] signature = Base64.decodeBase64(pieces[2]);
+    String baseString = JsonTokenUtil.toDotFormat(pieces[0], pieces[1]);
+
+    boolean sigVerified = false;
+    for (Verifier verifier : verifiers) {
+      AsciiStringVerifier asciiVerifier = new AsciiStringVerifier(verifier);
+      try {
+        asciiVerifier.verifySignature(baseString, signature);
+        sigVerified = true;
+        break;
+      } catch (SignatureException e) {
+        continue;
+      }
+    }
+    return sigVerified;
+  }
+
+  /**
+   * Verifies that a JSON Web Token is not expired.
+   *
+   * @param jsonToken the token to verify
+   * @param now the instant to use as point of reference for current time
+   * @return false if the token is expired, true otherwise
+   */
+  public boolean expirationIsValid(JsonToken jsonToken, Instant now) {
+    Instant expiration = jsonToken.getExpiration();
+    if ((expiration != null) && now.isAfter(expiration)) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Verifies that a JSON Web Token was issued in the past.
+   *
+   * @param jsonToken the token to verify
+   * @param now the instant to use as point of reference for current time
+   * @return false if the JWT's 'iat' is later than now, true otherwise
+   */
+  public boolean issuedAtIsValid(JsonToken jsonToken, Instant now) {
+    Instant issuedAt = jsonToken.getIssuedAt();
+    if ((issuedAt != null) && now.isBefore(issuedAt)) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * @param tokenString The original encoded representation of a JWT
+   * @return Three components of the JWT as an array of strings
+   */
+  private String[] splitTokenString(String tokenString) {
+    String[] pieces = tokenString.split(Pattern.quote(JsonTokenUtil.DELIMITER));
+    if (pieces.length != 3) {
+      throw new IllegalStateException("Expected JWT to have 3 segments separated by '" +
+          JsonTokenUtil.DELIMITER + "', but it has " + pieces.length + " segments");
+    }
+    return pieces;
+  }
+
+  /**
+   * Extracts the necessary information to look up verifiers.
+   *
+   * @param jsonToken the token to verify
+   * @return Signature algorithm, issuer, and keyId in an object
+   */
+  protected ProviderLookupData getLookupData(JsonToken jsonToken) {
+    JsonObject header = jsonToken.getHeader();
+    JsonElement keyIdJson = header.get(JsonToken.KEY_ID_HEADER);
+    String keyId = (keyIdJson == null) ? null : keyIdJson.getAsString();
+    SignatureAlgorithm sigAlg = jsonToken.getSignatureAlgorithm();
+
+    return new ProviderLookupData(sigAlg, jsonToken.getIssuer(), keyId);
+  }
+
+  /**
+   * Class that bundles up the necessary data to look up verifiers.
+   */
+  protected class ProviderLookupData {
```

Comment:
Might be worth using AutoValue here: https://github.com/google/auto/blob/master/value/userguide/index.md

Reply:
The ProviderLookupData only serves to encapsulates three values to pass along in a function. It is unlikely that we will need the hashcode or use equal and toString. Would it still be worth using AutoValue in these situations?

Reply:
There are other reasons to use AutoValue, see the slides here:
https://github.com/google/auto/blob/master/value/userguide/why.md

Reply:
Resolved. The `ProviderLookupData` class now uses AutoValue.